### PR TITLE
Bring-up of state trie hash verification function

### PR DESCRIPTION
**Pull request by shamb0**
Bring-up of state trie hash verification function.

Observing mismatch in hash value under FEVM context.

**Summary**

> Modified test vector with single transaction, common test vector used in both fevm & revm context.

- [mod-push0.json](https://github.com/shamb0/fevm-eth-compliance-test-trace/blob/main/GeneralStateTests/EIPTests/stEIP3855/TID-01-04-01-mod-push0.json#L1)

> Mismatch in computed state trie hash value under FEVM context

- [Complete Trace](https://github.com/shamb0/fevm-eth-compliance-test-trace/blob/main/GeneralStateTests/EIPTests/stEIP3855/TID-01-04-02-fevm.md#L5)

> Computed state trie hash value under revm context

- [Complete Trace](https://github.com/shamb0/fevm-eth-compliance-test-trace/blob/main/GeneralStateTests/EIPTests/stEIP3855/TID-01-04-03-revm.md#L1)

## Difference in state properties values, used for calculating state trie hash.

> Mismatch in bytecode hash

**FEVM context**

**revm context**

> Mismatch in slot status @0000000000000000000000000000000000000400

**FEVM context**

**revm context**

### CODE REVIEW

**Comment by shamb0** [comment]
CC @Stebalien, @vyzo

**Comment by Stebalien** [comment]
For the first case, can you print out the state trees from both cases? I.e., the slots so I can see where the difference is?

**Comment by Stebalien** [comment]
Ah... there aren't separate sub tests.

**Comment by Stebalien** [comment]
But yes, this is exactly what we're looking for.

One thing that's confusing me is that both FEVM and REVM agree on the state of `b94f5374fce5edbc8e2a8697c15331677e6ebf0b`, but the tests don't.

**Comment by Stebalien** [comment]
But I solved the bytecode hash issue.

https://github.com/shamb0/fevm-eth-compliance-test/blob/53648dcfb5921c26f9c20d87fcb5c6d030d2ed58/src/common/tester.rs#L384-L410

The bytecodehash as stored in the state is the hash digest, not a multihash. I.e., there's no need to wrap the hash digest in a multihahs, convert it to bytes, then take the first 32. Just take the entire digest:

https://github.com/shamb0/fevm-eth-compliance-test/blob/53648dcfb5921c26f9c20d87fcb5c6d030d2ed58/src/common/tester.rs#L390

If possible, I recommend that you use the EVM's builtin constructor.

**Comment by Stebalien** [comment]
https://github.com/shamb0/fevm-eth-compliance-test/pull/4

**Comment by Stebalien** [comment]
>     Mismatch in slot status @0000000000000000000000000000000000000400

It looks like you aren't inserting the slots from the "pre" section before running the test.

**Comment by shamb0** [comment]
Hello @Stebalien,

> Update-01 :: Quick fix to suppress the error `ExitCode::SYS_SENDER_INVALID` | "Send not from valid sender"

* Based on state of the condition flag, error "Send not from valid sender" is due to `is_placeholder_actor()` is not set.



* Indeed state of `is_placeholder_actor()` not set is valid, since fevm actor is created during runtime & it's not part of the manifest file.

* As a quick hack, Re-factored the logic as below to suppress the error. Could you please kindly suggest the better alternate to handle this.

https://github.com/shamb0/ref-fvm/blob/shamb0/mirror-master/fvm/src/executor/default.rs#L422

**Comment by shamb0** [comment]
> Update-02 :: Status on analysis of mismatch in state trie hash calculated value.

Here are some observation on the properties of `ActorState` for the address 
`0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b`

* Observation-01 :: Difference in `gas_used` between revm & fevm

| prop | fevm | revm |
| --- | --- | --- |
| base_fee | 0.00000000000000001 | TODO |
| initial_balance | 17592186044416 | 17592186044416 |
| balance_after_transaction | 17592186044416 |  17592184564106 |
| balance_diff | 0 | 1480310 |
| gas_used | 4367801 | 148031 |

from the [traces](https://github.com/shamb0/fevm-eth-compliance-test-trace/blob/main/GeneralStateTests/EIPTests/stEIP3855/TID-01-04-04-fevm.md), looks like `ActorState::deduct_funds()` is always passed with value zero, Can I have some clarity on this behavior ?



* Observation-02 :: Please correct me if I'm wrong, Looks like functionality of the property `ActorState::sequence` & `fil_actor_evm::State::nonce` are duplicates, but the values are in out-of sync.

**Comment by Stebalien** [comment]
> Send not from valid sender

You need to send messages from an _account_ or _placeholder_, not an EVM actor. The EVM doesn't distinguish between accounts and contracts, but we do.

To handle this, you likely need to check `AccountInfo::code`. If `code` is _empty_, create an account actor here:

https://github.com/shamb0/fevm-eth-compliance-test/blob/3d9c693092c46833b841e592f2a45b113cb0a762/src/statetest/executor.rs#L441

If it's non-empty, create an EVM actor as usual.

**Comment by Stebalien** [comment]
> from the [traces](https://github.com/shamb0/fevm-eth-compliance-test-trace/blob/main/GeneralStateTests/EIPTests/stEIP3855/TID-01-04-04-fevm.md), looks like ActorState::deduct_funds() is always passed with value zero, Can I have some clarity on this behavior ?

I'm not sure where you're getting those traces, I assume you've modified something?

I'm guessing that you're setting the base-fee to 0. In which case yes, we'll charge nothing for gas.


> Observation-02 :: Please correct me if I'm wrong, Looks like functionality of the property ActorState::sequence & fil_actor_evm::State::nonce are duplicates, but the values are in out-of sync.

That's a bit tricky:

1. `ActorState::sequence` is the _Filecoin_ nonce and is only used by accounts to order off-chain messages.
2. `fil_actor_evm::State::nonce` is the EVM nonce and is used in `CREATE` to derive the address.

`fil_actor_evm::State::nonce` exists because the EVM needs to update this nonce every time it creates a new contract, but actors aren't allowed to manually increment their Filecoin nonce (`ActorState::sequence`).

Short version: For accounts (ethereum-style accounts and native accounts), always look at `ActorState::sequence`. For everything else, ignore it.

**Comment by shamb0** [comment]
> To handle this, you likely need to check `AccountInfo::code`. If `code` is _empty_, create an account actor here:
> 
> https://github.com/shamb0/fevm-eth-compliance-test/blob/3d9c693092c46833b841e592f2a45b113cb0a762/src/statetest/executor.rs#L441
> 
> If it's non-empty, create an EVM actor as usual.

Looks like I'm missing some information piece, in managing place-holder address & fevm actor address. Could you please kindly review the latest integration ?

https://github.dev/shamb0/fevm-eth-compliance-test/blob/shamb0/verify-merkle-trie/src/statetest/executor.rs#L406

**Comment by shamb0** [comment]
> I'm guessing that you're setting the base-fee to 0. In which case yes, we'll charge nothing for gas.

Using the base-fee value from the test vector property. Can we use it straight or any transformation required ?

https://github.com/shamb0/fevm-eth-compliance-test-trace/blob/b333f5adef012b83ef287ded0bc4de1a8c77eafc/GeneralStateTests/EIPTests/stEIP3855/TID-01-04-01-mod-push0.json#LL22C2-L22C2

image

**Comment by Stebalien** [comment]
> Looks like I'm missing some information piece, in managing place-holder address & fevm actor address. Could you please kindly review the latest integration ?

- If `info.code` is empty, construct an EthAccount https://github.com/filecoin-project/builtin-actors/blob/89dc63dfc9473d7db42525bf59bb0e679d8c9c41/actors/ethaccount/src/lib.rs. Specifically, create an actor with empty state (`EMPTY_ARR_CID`) and with the EthAccount code CID (e.g., https://github.com/filecoin-project/ref-fvm/blob/456ab0d9783f4d486848af811c87d7a6bccb35e3/fvm/src/executor/default.rs#L427).
- Otherwise, create an EVM actor as usual.

**Comment by Stebalien** [comment]
> Using the base-fee value from the test vector property. Can we use it straight or any transformation required ?

What are you trying to test? The gas used by FEVM will always differ from the gas used by the EVM.

**Comment by shamb0** [comment]
> > Using the base-fee value from the test vector property. Can we use it straight or any transformation required ?
> 
> What are you trying to test? The gas used by FEVM will always differ from the gas used by the EVM.


Oops, then we can't use the state root hash to test against the `post::testcase::hash`, which is part of the test vector.



Computation of state root hash, has dependency on state of contract balance. (After executing the transaction use-cases, `gas_used` is 
deducted from the contract account balance).
